e3nn/o3/_linear.py

```diff
@@ -185,45 +297,48 @@ def _codegen_linear(
 
     out_list = []
 
-    instr = [
-        (i_in, i_out)
-        for i_in, (_, ir_in) in enumerate(irreps_in)
-        for i_out, (_, ir_out) in enumerate(irreps_out)
-        if ir_in == ir_out
-    ]
+    # TODO: ?? "fuse" instructions to make more efficient? I.e.:
+    # instead of (0, 0), (0, 1), etc., do one big einsum of input 0 with many weights,
+    # and then extract the relevant parts for summing the outputs.
 
-    for i_in, i_out in instr:
-        mul_ir_in = irreps_in[i_in]
-        mul_ir_out = irreps_out[i_out]
+    for ins in instructions:
+        mul_ir_in = irreps_in[ins.i_in]
+        mul_ir_out = irreps_out[ins.i_out]
 
         # Short-circut for empty irreps
         if mul_ir_in.dim == 0 or mul_ir_out.dim == 0:
             continue
 
         # Extract the weight from the flattened weight tensor
-        path_nweight = mul_ir_in.mul*mul_ir_out.mul
-        if len(instr) == 1:
+        path_nweight = prod(ins.path_shape)
+        if len(instructions) == 1:
+            # Avoid unnecessary view when there is only one weight
             w = ws
         else:
             w = ws[
                 :,
                 flat_weight_index:flat_weight_index + path_nweight
             ]
         w = w.reshape(
-            (() if shared_weights else (-1,)) + (mul_ir_in.mul, mul_ir_out.mul)
+            (() if shared_weights else (-1,)) + ins.path_shape
         )
         flat_weight_index += path_nweight
 
-        ein_out = torch.einsum(f"{z}uw,zui->zwi", w, x_list[i_in])
-        alpha = 1.0 / math.sqrt(mul_ir_in.mul * sum(1 if i_out_this == i_out else 0 for _, i_out_this in instr))
+        ein_out = torch.einsum(f"{z}uw,zui->zwi", w, x_list[ins.i_in])
+        alpha = 1.0 / math.sqrt(
+            mul_ir_in.mul * sum(
+                1 if other_ins.i_out == ins.i_out else 0
```

Comment:
Maybe I could make a remark in the doc that `Linear(10x0, 0)` is not equivalent to `Linear(2x0 + 8x0, 0)`

I seriously wonder if we should remove the `simplify` operations

Reply:
aha yes, this didn't occur to me — it should be removed and there should definitely be a note in the docs.

Maybe also more discussion somewhere in the docs in general that it can be very good to call `simplify()` yourself to get more efficient networks... maybe it is time to start an "FAQ" or a "Tips" page in the docs?

Reply:
Yes tips and/or faq is a good idea